### ISSUE TITLE
Using SAM and ASAM for RNNs can cause nan outputs

### ISSUE THREAD

**Issue by DaStapo** [davda54/sam#57]
Using SAM (and ASAM) for my RNN models eventually ends up in the models producing nan values and I'm wondering if there's a way to deal with that.

SAM clearly improves stability and generalization for my models, at least until the point when the nan values start to appear. I've tried reducing learning rates, batch sizes, dropout rates, gradient clipping and the SAM hyperparameters, yet it always explodes when the training loss is at about 80% of what it usually converges at.

Is there something I could change in the implementation, other than adjusting the rho hyperparameter? Would it be possible/reasonable to exclude the RNN layers from SAM's optimization (only using the base optimizer for RNNs)?

**Comment by DaStapo** [comment]
I assume that the way SAM ajdusts weights can make training to be more prone to exploding gradients. In my case it was an RNN architecture in which the values became too large when using too many frames.

I'd still love to hear if there's anything to adjust with SAM to help with exploding gradients, but there probably isn't anything of such, and neither is it a matter of this implementation, so I'm closing the issue.
